test: Bluetooth: Mesh: Test of the provisioning with OOB authentication

alxelax opened the pull request:
Tests check all possible OOB authentication methods and actions according to spec.
To pass all kinds of authentications the refactoring has been required since there were hidden dependencies between
provisioner behavior and settings for device capability. It was due to reusing the common authentication part for both
provisionee and provisioner. 
As an OOB interface to communicate between virtual devices the UDP client-server loop has been added.
As a final clean up the warnings in the scanner test have been fixed.

alxelax committed:
test: Bluetooth: Mesh: clean up warnings in Mesh scanner test

Clean up warnings in the BubbleSim Mesh scanner test.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax committed:
Bluetooth: Mesh: Refactoring provisioning to make all OOB auth working

The current implementation has hidden dependencies that break
OOB authentication if provisioner does not have the configured
input or output fields used for device capabilities.
It didn't allow to pass several OOB authentication cases.
After refactoring provisioner behavior is independent to
provisionee settings.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_provision.c:
```diff
@@ -3,9 +3,13 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
+#include <stdio.h>
```
Is this a left over from debugging?

alxelax replied:
yep, removed

trond-snekvik commented on tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/provision/pb_adv_oob_auth.sh:
```diff
@@ -0,0 +1,11 @@
+	RunTest mesh_provision_pb_adv_oob_auth \
+		prov_device_pb_adv_oob_auth \
+		prov_provisioner_pb_adv_oob_auth -- -argstest test-step=$i
+done
```
Wouldn't it be faster/simpler to perform the loop within the test scenario?

alxelax replied:
There is no way to unprovision device.

trond-snekvik replied:
`bt_mesh_reset()`

alxelax replied:
it doesn't seem to work 
```
d_01: @00:00:00.002648  [00:00:00.002,624] <wrn> bt_ecc: Callback already registered
d_01: @00:00:00.002648  [00:00:00.002,624] <err> bt_mesh_prov: Failed to generate public key (-114)
```
not possible to start mesh again because of public key error

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/oob_client.c:
```diff
@@ -0,0 +1,54 @@
+	close(sockfd);
+
+	return err;
+}
```
@trond-snekvik There is already :): https://github.com/BabbleSim/base/blob/master/libPhyComv1/src/bs_pc_backchannel.c#L7

One thing you should be aware of, when you build this kind of solution (like in this PR), with just 1 of something (one socket, one file, one whathaveyou) is that you cannot run multiple simulations in parallel. Not with different sim_ids, and not by different users in the same machine, or sharing the same drive/folder. That can be a significant problem, for ex., if you want to run things in parallel for speed,  for CI systems, or so. The basic solution for this in the libraries provided with bsim is that results/dump/intermediate files are all kept in the results/<sim_id>/ folder. And any temporary "files" are kept in /tmp/bs_<user>/<sim_id>/. You don't need to use that if you don't want to. But if you don't, you should be mindful of not shooting your foot to badly.

alxelax replied:
Really I do not want the universe spaceship here. I added the OOB interface according to Mesh spec usage. This is p2p communication to provide simple data and closes the socket immediately. Do not use it for BubbleSim debug purposes it will leave your foots intact :)

aescolar replied:
@alxelax What happens if 2 users in the same computer run the same simulation at the same time, or 2 simulations with 2 different sim_ids happen to use this OOB mechanism at the same time? :)

alxelax replied:
@aescolar That is quite obvious since UDP is the transport layer. All port listeners will receive the same data. The problem that you described is not unique and has been solved a long time ago by creating an extended ISO model. There are Session and Presentation layers to solve that. The question is how possible is the situation in real life for these tests that you described here? Are there real use cases?

trond-snekvik replied:
It seems to me like the existing backchannel solution would be the right approach here, even if the novel solution in this file works properly. 

@aescolar is there some sample or existing application that uses this, that we can use as a reference? Now that I know this feature exists, I'd like to use it actively in tests. I suppose it could even be integrated into the test framework with a standard channel for creating sync points.

alxelax replied:
Ok. I'll change my implementation because this is a code owner request. However, I still think it is not a good idea. Trying to fix what is not broken basing on strange assumptions.

aescolar replied:
@trond-snekvik 
> is there some sample or existing application that uses this, that we can use as a reference?

Unfortunately nothing open sourced. I would hope that the description/comments in that file (see each function comment) would be good enough to get started. The functionality the backchannels provide is quite simple.

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_mesh/CMakeLists.txt:
```diff
@@ -29,10 +31,12 @@ else()
  )
 endif()
 
+zephyr_compile_options(-Werror)
```
@trond-snekvik It is a bit too much to do this here. As one may want (say while debugging) to do something sloppy that causes warnings, and not have the compiler prevent you. In twister, it is twister the one raising warnings to errors, to ensure no warnings get pass.
About your question: `compile.sh` was done dropping the output unless there is errors to be silent (the "silence is golden" Unix rule), to not pollute your console, and to not pollute the CI output. It would be possible to add an option for compile.sh that sets `-Werror`. Say either just adding `cmake_args="-DEXTRA_CFLAGS=-Werror"` to that particular app line in compile.sh, or adding a new option overall for compile.sh that adds that same flag. (Note that compile.sh was born just as a script for CI )

alxelax replied:
The main mode of work is the normal test running. I do not see any reason to hide warnings (that might be real bugs. Good example, where I made a bug and had to fix it emergently: https://github.com/zephyrproject-rtos/zephyr/pull/37079)
This is quite unfriendly to the new users. They do not know hidden assumptions and merge warnings and bugs in the main repo. It happens again and again. Seems Unix rule doesn't work well here.

alxelax replied:
I'll revert my changes related to this. Submited bug for BubbleSim tests build system.
https://github.com/zephyrproject-rtos/zephyr/issues/38303

alxelax commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -107,10 +107,98 @@ static bt_mesh_input_action_t input_action(uint8_t action)
+	bt_rand(str, size);
+
+	/* Normalize to '0' .. '9' & 'A' .. 'Z' */
+	for (uint8_t i = 0U; i < size; i++) {
+		str[i] %= 36;
+		if (str[i] < 10) {
+			str[i] += '0';
+		} else {
+			str[i] += 'A' - 10;
+		}
+	}
```
It is the existing algorithm that I moved in a separate function to not encumber huge auth switch case. Do you want me to refactor it here?

trond-snekvik replied:
Right, sorry. We could do this separately, it's not directly relevant to this PR then.

alxelax replied:
Applied your changes since the current algorithm generates unreadable symbols if change type 'unsigned char' to 'char'. Otherwise, it requires unnecessary types convertions that @PavelVPV  suggested to remove.

alxelax commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -107,10 +107,98 @@ static bt_mesh_input_action_t input_action(uint8_t action)
+
+	bt_rand(&num, sizeof(num));
+
+	for (uint8_t i = 1; i < size + 1; i++) {
+		divider *= 10;
+	}
```
The lookup table looks awful(digit with 16th zeroes :) ) when you try to add full spec supports. I guess it is the reason for the artificial limitation of authentication value wide. The current implementation doesn't allow 16 bytes authentications but Mesh spec does. This should be fixed to follow spec 100%

alxelax replied:
If you insist I can put lookup table back but it should go out in any case later.

trond-snekvik replied:
In section 5.4.1.3, the Provisioning Start message limits the provisioning size to 8 for input and output OOB. As far as I can tell, it's just the static OOB that permits a 16 byte OOB value

alxelax replied:
Ok. I missed this table in spec.

alxelax commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -107,10 +107,98 @@ static bt_mesh_input_action_t input_action(uint8_t action)
+		 * and vibrate should be a random integer
+		 * between 0 and 10^size, *exclusive*:
+		 */
+		num = (num % (divider - 1)) + 1;
+	} else {
+		num %= divider;
```
As for string I just moved the existing algorithm into a separate function. Didn't try to improve randomization. I suggest to do it in separate PR. This PR fixes OOB authentication. It wasn't possible to pass a couple of them at all.

alxelax commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -107,10 +107,98 @@ static bt_mesh_input_action_t input_action(uint8_t action)
+	return num;
+}
+
+int bt_mesh_prov_auth(bool is_provisioner, uint8_t method, uint8_t action, uint8_t size)
```
This is really bad practice. This mimics what it is not really. Especially for output and input actions that are not abstract digit and string. Convert `BT_MESH_BLINK` to `BT_MESH_PUSH` and `BT_MESH_TWIST` to `BT_MESH_BEEP` etc etc. It has been implemented before and it was difficult to understand why `BT_MESH_BLINK` worked but `BT_MESH_PUSH` didn't and why the provisioner required size setting for method that device doesn't use. Less mimic magic.
Moving  `check_..._auth` outside prov.c causes one more huge `switch .. case` that will encumber provisioning device implementation with variety size checking. 
I'd prefer to leave what I refactored.

trond-snekvik replied:
I don't fully agree that it's bad practice, as long as we're clear on this function's purpose. If its purpose is "execute the authentication procedure for this provisioning session", I agree that it's odd to feed it the wrong values. If its purpose is "run this authentication procedure for me", as called by the provisioner and provisioning device, it makes perfect sense that the provisioner has its own logic to feed the algorithm, IMO. 

I favor the second interpretation, as the implementation in this patch is leaking role specific details into this module, which doesn't happen in other parts of this file.

alxelax replied:
As I've already mentioned the authentication procedure is not fully symmetrical. There are more requirements to the provisioning device rather than to provisioner. It is not possible to use the same implementation as-is for both sides (btw. trying to use it caused all bugs that I fixed here). The number of authentication methods is not symmetrical depending on direction. When you mimic 3 output auth to 2 input auth it is not possible unequivocally to understand which output OOB is exactly used. Furthermore, the device requires more strict auth OOB sizes checking. Using "run this authentication procedure for me" approach requires exposing this functionality outside of prov.c
I'd agree with your design idea but it requires even more refactoring. Both sides provisioning device and provisioner should have its own switch case regarding OOB methods and call simple functions from prov.c like `generate number for blinking`, `generate general number`, `generate string`  etc.

alxelax committed:
test: Bluetooth: Mesh: add provisioning with OOB auth tests

BubbleSim tests that check provisioning with all possible (from spec)
OOB authentications

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

PavelVPV reviewed (commented):
Just couple of minor comments

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_provision.c:
```diff
@@ -130,6 +298,60 @@ static void test_provisioner_pb_adv_no_oob(void)
+
+	oob_auth_set();
+
+	oob_channel_id = bs_open_back_channel(0, (unsigned[]){1}, (unsigned[]){0}, 1);
```
A sidecomment: Maybe comment in the code here (and in the test script) that this test assumes to be running in device 0, and expects the other side of the test in device 1, and viceversa in the other test.

trond-snekvik commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -131,89 +219,70 @@ int bt_mesh_prov_auth(uint8_t method, uint8_t action, uint8_t size)
-			return -EINVAL;
+		if (!is_provisioner) {
+			err = check_input_auth(input, size);
+			if (err) {
+				return err;
+			}
 		}
 
-		if (size > bt_mesh_prov->input_size) {
-			return -EINVAL;
+		if (!is_provisioner) {
```
Combine into one `if`

alxelax replied:
looks like you requested to do this change 5 days ago https://github.com/zephyrproject-rtos/zephyr/pull/38205#discussion_r700189030
Now you ask to combine back. Didn't I understand anything?

alxelax replied:
Never mind, got it. will fix

trond-snekvik commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -131,89 +219,70 @@ int bt_mesh_prov_auth(uint8_t method, uint8_t action, uint8_t size)
-
-		if (size > bt_mesh_prov->output_size) {
-			return -EINVAL;
+		if (!is_provisioner) {
+			err = check_output_auth(output, size);
+			if (err) {
+				return err;
+			}
 		}
```
Move below?

trond-snekvik commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -239,11 +308,15 @@ int bt_mesh_input_string(const char *str)
 	}
 
-	strncpy((char *)bt_mesh_prov_link.auth, str, bt_mesh_prov->input_size);
+	strncpy((char *)bt_mesh_prov_link.auth, str, strlen(str));
```
This could just be `strcpy()`

trond-snekvik commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -239,11 +308,15 @@ int bt_mesh_input_string(const char *str)
 {
 	BT_DBG("%s", log_strdup(str));
 
+	if (strlen(str) > PROV_IO_OOB_SIZE_MAX) {
```
Should this also check against `bt_mesh_prov->input_size`?

alxelax replied:
Nop. This will not work since the function is used for both provisionee and provisioner. Provisioner will use it for output string OOB. More correctly to use here `bt_mesh_prov_link.oob_size`. However, I see oob fields are not set for provisionee site. Will fix.

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_provision.c:
```diff
@@ -14,20 +16,56 @@
+
 static ATOMIC_DEFINE(flags, TEST_FLAGS);
+
+extern void bs_clean_back_channels(void);
```
Why not just include `bs_pc_backchannel.h`?

alxelax replied:
no path to the header

alxelax replied:
in principle, it is a bad idea to do source code dependable on the third part repository

PavelVPV replied:
That seems strange as the path is added here: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/posix/nrf52_bsim/CMakeLists.txt#L37
And here `bs_cmd_line.h` is included without any issues: https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_persistence.c#L14

alxelax replied:
I see. It works only with `system` header including. I'll add it but it looks weird

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_provision.c:
```diff
@@ -87,14 +130,118 @@ static void prov_reset(void)
+	LOG_INF("OOB Number: %u", number);
+	snprintf(str, sizeof(str), "%d", number);
+
+	bs_bc_send_msg(*oob_channel_id, (uint8_t *)str, strlen(str));
```
A sidenote: If you want, you can send directly `number` without converting it to a string (and converting it back on the other side). You should be able to do:
`bs_bc_send_msg(*oob_channel_id, (uint8_t *)number, sizeof(number));`
